File: 3.x.x/crates-utopia/emails/src/canonicals/walla.rs
use crate::canonicals::provider::{to_lower_case, Canonical, Provider};
use crate::EmailError;

/// Walla canonicalization (PHP `Walla`) - case + domain only.
#[derive(Debug, Clone, Copy, Default)]
pub struct Walla;

const SUPPORTED_DOMAINS: &[&str] = &["walla.co.il", "walla.com"];
const CANONICAL_DOMAIN: &str = "walla.co.il";

impl Provider for Walla {
    fn supports(&self, domain: &str) -> bool {
        SUPPORTED_DOMAINS.contains(&domain)
    }

    fn get_canonical(&self, local: &str, _domain: &str) -> Result<Canonical, EmailError> {
        Ok(Canonical {
            local: to_lower_case(local),
            domain: CANONICAL_DOMAIN.to_string(),
        })
    }

    fn get_canonical_domain(&self) -> &'static str {
        CANONICAL_DOMAIN
    }

    fn get_supported_domains(&self) -> &'static [&'static str] {
        SUPPORTED_DOMAINS
    }
}

File: packages/emails/src/Canonicals/Providers/Walla.php
<?php

namespace Utopia\Emails\Canonicals\Providers;

use Utopia\Emails\Canonicals\Provider;

/**
 * Walla
 *
 * Handles Walla email normalization
 * - Supports both walla.co.il and walla.com domains
 * - Normalizes to walla.co.il domain
 * - Preserves all characters in local part (no plus addressing or dot removal)
 */
class Walla extends Provider
{
    private const SUPPORTED_DOMAINS = ['walla.co.il', 'walla.com'];

    private const CANONICAL_DOMAIN = 'walla.co.il';

    public function supports(string $domain): bool
    {
        return in_array($domain, self::SUPPORTED_DOMAINS, true);
    }

    public function getCanonical(string $local, string $domain): array
    {
        // Convert to lowercase
        $normalizedLocal = $this->toLowerCase($local);

        return [
            'local' => $normalizedLocal,
            'domain' => self::CANONICAL_DOMAIN,
        ];
    }

    public function getCanonicalDomain(): string
    {
        return self::CANONICAL_DOMAIN;
    }

    public function getSupportedDomains(): array
    {
        return self::SUPPORTED_DOMAINS;
    }
}
